include/RAJA/exec-cuda/MemUtils_CUDA.hxx

```diff
@@ -62,57 +62,282 @@
+  GridSizeType retiredBlocks;
 };
 
+
 /*!
-*************************************************************************
-*
-* Return next available valid reduction id, or complain and exit if
-* no valid id is available.
-*
-*************************************************************************
-*/
+ ******************************************************************************
+ *
+ * \brief Get a valid reduction id, or complain and exit if no valid id is 
+ *        available.
+ *
+ * \return int the next available valid reduction id.
+ *
+ ******************************************************************************
+ */
 int getCudaReductionId();
 
 /*!
-*************************************************************************
-*
-* Release given redution id and make inactive.
-*
-*************************************************************************
-*/
+ ******************************************************************************
+ *
+ * \brief Release given reduction id and make inactive.
+ *
+ ******************************************************************************
+ */
 void releaseCudaReductionId(int id);
 
-CudaReductionBlockTallyType* getCudaReductionTallyBlock(int id);
+/*!
+ ******************************************************************************
+ *
+ * \brief Get tally block for reducer object with given id.
+ *
+ * \param[out] host_tally pointer to host tally cache slot.
+ * \param[out] device_tally pointer to device tally slot.
+ *
+ * NOTE: Tally Block size will be:
+ *
+ *          sizeof(CudaReductionDummyTallyType) * RAJA_MAX_REDUCE_VARS
+ *
+ *       For each reducer object, we want a chunk of device memory that
+ *       holds the reduced value and a small number of anciliary variables.
+ *
+ ******************************************************************************
+ */
+void getCudaReductionTallyBlock(int id, void** host_tally, void** device_tally);
+
+/*!
+ ******************************************************************************
+ *
+ * \brief Release tally block for reducer object with given id.
+ * 
+ ******************************************************************************
+ */
+void releaseCudaReductionTallyBlock(int id);
+
+/*!
+ ******************************************************************************
+ *
+ * \brief Sets up state variales before the loop body is copied and the kernel
+ *        is launched.
+ *
+ ******************************************************************************
+ */
+void beforeCudaKernelLaunch();
+
+/*!
+ ******************************************************************************
+ *
+ * \brief Resets state variables after kernel launch.
+ *
+ ******************************************************************************
+ */ 
+void afterCudaKernelLaunch();
+
+/*!
+ ******************************************************************************
+ *
+ * \brief Updates host tally cache for read by reduction varaible with id.
+ *
```

Comment:
Should comment both of these or note that there are sync and non-sync versions of these and add '///' on line 310 so second one gets documented.

Would it be better to make these templates with the reduction policy as the template argument and explicitly specialize for each case? This would help avoid calling the wrong one erroneously.